lib/matplotlib/axes/_base.py

```diff
@@ -265,21 +253,18 @@ def _getdefaults(self, ignore, *kwargs):
             default_dict = {}
         return default_dict
 
-    def _setdefaults(self, defaults, *kwargs):
+    def _setdefaults(self, defaults, kw):
         """
-        Given a defaults dictionary, and any other dictionaries,
-        update those other dictionaries with information in defaults if
-        none of the other dictionaries contains that information.
-
+        Add to *kw* the entries in *default* that are absent or set to None in
+        *kw*.
         """
         for k in defaults:
-            if all(kw.get(k, None) is None for kw in kwargs):
-                for kw in kwargs:
-                    kw[k] = defaults[k]
+            if kw.get(k, None) is None:
```

Comment:
thats
~~~
kw.update({k: v
           for k, v in defaults.items()
           if kw.get(k, None) is None})
~~~

Not sure if that is more readable

Reply:
You can bet that if anything is writable as a comprehension I will likely have tried doing so.  In this case I thought the appearance of kw twice, outside and inside, hinders legibility.

Reply:
The double appearance in the expression is inherent to the problem. Essentially what we have to to (and what the expression also says):

Update kw
with keys and values from defaults
where the key is not in kw or maps to None

The more I think about it, the more I like the comprehension :smile:.

Note also that `kw.get(k, None)` is the same as `kw.get(k)`.

Reply:
Actually I thought the clearest was way to just drop out the Nones first, ie
```
    def _getdefaults(self, ignore, kw):
        """
        If some keys in the property cycle (excluding those in the set
        *ignore*) are absent or set to None in the dict *kw*, return a copy
        of the next entry in the property cycle, excluding keys in *ignore*.
        Otherwise, don't advance the property cycle, and return an empty dict.
        """
        prop_keys = self._prop_keys - ignore
        keys_no_none = {k for k, v in kw.items() if v is not None}
        if prop_keys - keys_no_none:  # Some keys in prop_keys are not in kw.
            # Copy the default dict to avoid modifying the reference held by
            # the cycler.
            default_dict = next(self.prop_cycler).copy()
            for p in ignore:
                default_dict.pop(p, None)
        else:
            default_dict = {}
        return default_dict
```
and
```
    def _setdefaults(self, defaults, kw):
        """
        Add to the dict *kw* the entries in the dict *default* that are absent
        or set to None in *kw*.
        """
        kw_no_none = {k: v for k, v in kw.items() if v is not None}
        kw.clear()
        kw.update(defaults)
        kw.update(kw_no_none)
```
How does that look to you?

Reply:
Actually, `self._setdefaults(default_dict, kw)` is basically `kw = {**default_dict, **kw}`. So if we can replace the kw instance by a new one (which I suppose we can), we can get rid of the function alltogether. Or you could use a `collections.ChainMap`.

Reply:
Better but still a bit verbose. It's less efficient than the comprehension, but with few keywords that probably doesn't matter.

What about replacing the function with:

~~~
def _no_none(kw):
    return {k: v for k, v in kw.items() if v is not None}

kw = {**defaults, **_no_none(kw)}
~~~

Reply:
Indeed, introducing the _no_none helper makes the whole thing much simpler.  Pushed a new version.

Reply:
Actually, that doesn't work: we don't want to drop all the None entries in kw, *only those that are present in defaults*.  Those that are not in defaults should be kept.
(This is needed e.g. for the `closed` kwarg to `fill`.)

Returning to the previous version...